src/main/java/org/springframework/data/repository/reactive/RxJava2CrudRepository.java

```diff
@@ -0,0 +1,170 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.data.repository.reactive;
+
+import io.reactivex.Completable;
+import io.reactivex.Flowable;
+import io.reactivex.Maybe;
+import io.reactivex.Single;
+
+import java.io.Serializable;
+
+import org.springframework.data.repository.NoRepositoryBean;
+import org.springframework.data.repository.Repository;
+
+/**
+ * Interface for generic CRUD operations on a repository for a specific type. This repository follows reactive paradigms
+ * and uses RxJava 2 types.
+ *
+ * @author Mark Paluch
+ * @since 2.0
+ * @see Single
+ * @see Flowable
+ * @see Completable
+ */
+@NoRepositoryBean
+public interface RxJava2CrudRepository<T, ID extends Serializable> extends Repository<T, ID> {
+
+	/**
+	 * Saves a given entity. Use the returned instance for further operations as the save operation might have changed the
+	 * entity instance completely.
+	 *
+	 * @param entity must not be {@literal null}.
+	 * @return the saved entity.
+	 */
+	<S extends T> Single<S> save(S entity);
+
+	/**
+	 * Saves all given entities.
+	 *
+	 * @param entities must not be {@literal null}.
+	 * @return the saved entities.
+	 * @throws IllegalArgumentException in case the given entity is {@literal null}.
+	 */
+	<S extends T> Flowable<S> save(Iterable<S> entities);
+
+	/**
+	 * Saves all given entities.
+	 *
+	 * @param entityStream must not be {@literal null}.
+	 * @return the saved entities.
+	 * @throws IllegalArgumentException in case the given {@code Publisher} is {@literal null}.
+	 */
+	<S extends T> Flowable<S> save(Flowable<S> entityStream);
+
+	/**
+	 * Retrieves an entity by its id.
+	 *
+	 * @param id must not be {@literal null}.
+	 * @return the entity with the given id or {@link Maybe#empty()} if none found.
+	 * @throws IllegalArgumentException if {@code id} is {@literal null}.
+	 */
+	Maybe<T> findOne(ID id);
+
+	/**
+	 * Retrieves an entity by its id supplied by a {@link Single}.
+	 *
+	 * @param id must not be {@literal null}.
+	 * @return the entity with the given id or {@link Maybe#empty()} if none found.
+	 * @throws IllegalArgumentException if {@code id} is {@literal null}.
+	 */
+	Maybe<T> findOne(Single<ID> id);
+
+	/**
+	 * Returns whether an entity with the given id exists.
+	 *
+	 * @param id must not be {@literal null}.
+	 * @return {@literal true} if an entity with the given id exists, {@literal false} otherwise.
+	 * @throws IllegalArgumentException if {@code id} is {@literal null}.
+	 */
+	Single<Boolean> exists(ID id);
+
+	/**
+	 * Returns whether an entity with the given id exists supplied by a {@link Single}.
```

Comment:
rewording as `Returns whether an entity with the given id, supplied by a {@link Single}, exists` would be a tiny bit clearer IMHO